### ISSUE TITLE
Change Request: Export Config type

### ISSUE THREAD

**Issue by rchl** [eslint/eslint#18911]
### ESLint version

v9.10.0

### What problem do you want to solve?

I'd like to create a shared config that exports an array of (flat) eslint configs. Current problem with that seems to be that I can't use the [`Config` type](https://github.com/eslint/eslint/blob/2de5742682ec45e24dca9ca7faaa45330497fca9/lib/types/index.d.ts#L1247) from the `Linter` namespace since it's not exported. I would like to be able to import that type so that I can create a type declaration file like.

```ts
import { Linter } from 'eslint';

export default Linter.Config[];
```

(Note that this is not an issue when using JSDoc since somehow it's allowed to reference non-exported types from within JSDoc - `/** @type {import('eslint').Linter.Config[]} */` works. But this doesn't help when wanting to include public type declaration in a shared config.)

### What do you think is the correct solution?

Export `Config` so that I can reference it on the imported `Linter` namespace.

Potentially other types should also be exported.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

I've noticed that `typescript-eslint` for example [re-creates Config types](https://github.com/typescript-eslint/typescript-eslint/blob/af92611f53d21243656d5588e9e6a5fde1ab4d51/packages/utils/src/ts-eslint/Config.ts#L136-L150) which seems a bit problematic as then the configs it exports are not compatible type-wise with eslint's own `Config` types.

**Comment by nzakas** [comment]
Can you provide a reproduction case that shows the problem you're having?

I feel like I've seen people use `Linter.Config` elsewhere, so it would be helpful to see a case where it's not working to be able to evaluate.

**Comment by rchl** [comment]
Just try the declaration file like I've shown in my initial post and see if you can reference the `Linter.Config` type. If not then that's a problem because that means one can't create sharable config with proper public types.

In project-local eslint config file people are using config with .js/.mjs/.cjs extension so then they can use JSDoc type annotations which works (I haven't looked into why it's allowed to access non-public namespace fields in JSDoc but not TS).

**Comment by rchl** [comment]
Sure, I understand.

It's hard to show the "full flow" of the issue since it would require one package with shared eslint config and another that consumes it. That said, here is how an example shared config could look like, including exported types: https://stackblitz.com/edit/vitejs-vite-uzjytc?file=src%2Findex.d.ts

As previously mentioned, that won't currently work because `Linter.Config` type is not public.

With a shared config like that, then there would be another package that consumes it that would have an eslint config like:

```js
import mySharedConfig from 'my-shared-eslint-config';

/** @type {import('eslint').Linter.Config[]} */
export default [
  ...mySharedConfig,
  // Other rules...
];
```

And to re-iterate: packages like `typescript-eslint` expose their own invented types from their package. Maybe because eslint's ones are not publicly exposed. So importing their package like that causes issues as their types are not compatible with `Linter.Config[]` (even if runtime linting all works fine)

**Comment by iNewLegend** [comment]
@rchl Did you find some temporal solution?

**Comment by rchl** [comment]
There is always a way to cast types or ignore type errors with `// @ts-expect-error`. Those are hacky workarounds of course.

**Comment by nzakas** [comment]
@rchl we really need something that actually shows the error you're seeing, not just an example that might trigger it.

The StackBlitz you provided doesn't show any errors.

**Comment by rchl** [comment]
Because stackbliz can't do proper type checking. You need to download the project locally, install dependencies and then you'll see the type issue.

**Comment by rchl** [comment]
It seems to work so I don't know what I was seeing. Sorry for wasting your time.

**Comment by nzakas** [comment]
No worries. Glad you figured it out.
